Library/Source/NativeXr.cpp

```diff
@@ -93,11 +95,46 @@ namespace
         worldSpaceTransform[14] = pos.Z;
         worldSpaceTransform[15] = 1.f;
 
-        // Invert to get the view space transform.
-        std::array<float, 16> viewSpaceTransform{};
-        bx::mtxInverse(viewSpaceTransform.data(), worldSpaceTransform.data());
+        if (viewSpace)
+        {
+            // Invert to get the view space transform.
+            std::array<float, 16> viewSpaceTransform{};
+            bx::mtxInverse(viewSpaceTransform.data(), worldSpaceTransform.data());
+
+            return viewSpaceTransform;
+        }
+        else
+        {
+            return worldSpaceTransform;
+        }
+    }
 
-        return viewSpaceTransform;
+    void SetXRInputSourceSpaces(Napi::Object& jsInputSource, xr::System::Session::Frame::InputSource& inputSource)
+    {
+        auto env = jsInputSource.Env();
+        jsInputSource.Set("targetRaySpace", Napi::External<decltype(inputSource.AimSpace)>::New(env, &inputSource.AimSpace));
+        jsInputSource.Set("gripSpace", Napi::External<decltype(inputSource.GripSpace)>::New(env, &inputSource.GripSpace));
+    }
+
+    Napi::ObjectReference CreateXRInputSource(xr::System::Session::Frame::InputSource& inputSource, Napi::Env& env)
+    {
+        constexpr std::array<const char*, 2> HANDEDNESS_STRINGS
+        {
+            "left",
+            "right"
+        };
+        constexpr char* TARGET_RAY_MODE{ "tracked-pointer" };
+
+        auto jsInputSource = Napi::Object::New(env);
+        jsInputSource.Set("handedness", Napi::String::New(env, HANDEDNESS_STRINGS[static_cast<size_t>(inputSource.Handedness)]));
+        jsInputSource.Set("targetRayMode", TARGET_RAY_MODE);
+        SetXRInputSourceSpaces(jsInputSource, inputSource);
+
+        // auto profiles = Napi::Array::New(env);
+        // Napi::Value string = Napi::String::New(env, "generic-trigger-squeeze-touchpad-thumbstick");
+        // profiles.Set(0, string);
```

Comment:
what's this?

Reply:
WIP feature from when I was waiting to consume a new Babylon.js; it's how Babylon will eventually know what controller model to display. Doesn't show it yet, but the implementation doesn't hurt anything and needs to be there anyway, so I just went ahead and uncommented this and finished it in my most recent check-in.